### PR TITLE
feat(dialog): add onDismiss callback to support "soft dismiss" functionality

### LINKED ISSUE #861: Soft dismiss callback on Dialog
Often times "dialogs" or "lightboxes" have soft-dismiss options. I'd like to request a similar feature for Dialog. A new prop, "onDismiss" could be passed in that gets called when a user clicks on the modal overlay. Or a similar functionality that gives the consuming component a way to know when a user has clicked outside the content.

### PR DESCRIPTION
**Pull request by chrisdholt**
<body>
Adds a new prop of `onDismiss`. When passed, the callback will be fired in the following situations:
- When a modal dialog is open and the modal overlay is clicked
- When a modal or non-modal dialog is open and the escape key is pressed

closes #861 
<footer>
 

For [details on formatting](https://github.com/Microsoft/fast-dna/wiki/pull-request-guidance) pull requests.

### CODE REVIEW

File: packages/fast-components-react-base/src/dialog/dialog.props.ts [hunk 1]
```diff
@@ -40,13 +40,20 @@ export interface IDialogHandledProps {
      */
     modal?: boolean;
 
+    /**
+     * The onDismiss callback
+     * Callback is registered on click of the modal or when escape key is pressed
```
Comment (nalogiudice) [inline packages/fast-components-react-base/src/dialog/dialog.props.ts]: "...on click of modal" means when modal overlay is clicked, correct?

File: packages/fast-components-react-base/src/dialog/dialog.tsx [hunk 2]
```diff
@@ -50,6 +53,15 @@ class Dialog extends Foundation<IDialogHandledProps & IManagedClasses<IDialogCla
         );
     }
 
+    /**
+     * React life-cycle method
+     */
+    public componentDidMount(): void {
+        if (canUseDOM() && this.props.onDismiss) {
+            window.addEventListener("keydown", this.onDialogKeyboardDismiss);
```
Comment (nalogiudice) [inline packages/fast-components-react-base/src/dialog/dialog.tsx]: Don't we need to remove the event listener on unMount?

Comment (chrisdholt) [reply packages/fast-components-react-base/src/dialog/dialog.tsx]: yep - updated.

File: packages/fast-components-react-base/src/dialog/dialog.props.ts (hunk 1 above)
Comment (chrisdholt) [reply packages/fast-components-react-base/src/dialog/dialog.props.ts]: yes. Updated with more detail.

File: packages/fast-components-react-base/src/dialog/dialog.tsx [hunk 3]
```diff
@@ -50,6 +53,35 @@ class Dialog extends Foundation<IDialogHandledProps & IManagedClasses<IDialogCla
...
+        }
+    }
+
+    /**
+     * React life-cycle method
+     */
+    public componentDidUpdate(prevProps: Partial<IDialogHandledProps>): void {
+        if (canUseDOM() && !prevProps.onDismiss && this.props.onDismiss) {
```
[the remark this answers is not in the document]
Comment (chrisdholt) [reply packages/fast-components-react-base/src/dialog/dialog.tsx]: agreed and updated

File: packages/fast-components-react-base/src/dialog/dialog.tsx [hunk 4]
```diff
@@ -68,11 +80,26 @@ class Dialog extends Foundation<IDialogHandledProps & IManagedClasses<IDialogCla
...
+
+    private onClickDismiss = (event: React.MouseEvent): void => {
+        if (!this.props.onDismiss || !this.props.visible) {
+            return;
+        }
+
+        this.props.onDismiss(event);
+    }
+
+    private onDialogKeyboardDismiss = (event: KeyboardEvent): void => {
```
[the remark this answers is not in the document]
Comment (chrisdholt) [reply packages/fast-components-react-base/src/dialog/dialog.tsx]: updated

File: packages/fast-components-react-base/src/dialog/dialog.tsx [hunk 5]
```diff
@@ -68,11 +102,24 @@ class Dialog extends Foundation<IDialogHandledProps & IManagedClasses<IDialogCla
...
                 className={get(this.props, "managedClasses.dialog_modalOverlay")}
+                onClick={this.handleOverlayClick}
                 role={"presentation"}
                 tabIndex={-1}
             />
         );
     }
+
+    private handleOverlayClick = (event: React.MouseEvent): void => {
+        if (this.props.onDismiss && this.props.visible) {
```
Comment (nicholasrice) [inline packages/fast-components-react-base/src/dialog/dialog.tsx]: you should also make sure `this.props.onDismiss` is a *function*, not just a truthy value. `typeof this.props.onDismiss === "function"`

**Comment by awentzel** [comment]
I don't believe that modal dialogs should support 'escape' keypress to trigger 'onDismiss'. For example 'Cancel' or 'Accpet'.  Now on 'Cancel', it would make sense to trigger 'onDismiss'. So if there is a cancel button I could see that 'escape' would be acceptable.  

However, what if the two buttons were 'Accept' or 'Reject' and neither was defined by 'cancel' action.  You wouldn't want 'escaping' because it wouldn't have a valid return value which is required.

Is this an optional?  What is the default?

**Pull request merged**